### ISSUE TITLE
Are the OpenUSD libraries when compiled with python support thread safe?

### ISSUE THREAD

**Issue by ggarra13** [PixarAnimationStudios/USD#3118]
### Description of Issue
I am using libusd, etc. without python support in my viewer.  When compiled without python support, the build succeeds and opening a stage from any C++ thread works fine.
However, when I compile it with python support, the build succeeds but opening a stage from any C++ thread leads to hanging.

So the quick question is whether the USD libraries are thread safe when compiled with python support?

### Steps to Reproduce
1. See above and below (build flags) for the question.

### System Information (OS, Hardware)
Probably all platforms, but tested on:

- Ubuntu 22.04.4 LTS
- macOS Sonoma

### Package Versions
USD v24.05

### Build Flags

Works with threading support:

`python3.11 build_scripts/build_usd.py --no-python --no-examples --no-tutorials --no-tools --no-usdview --generator Ninja --verbose`

Does not work with threading support:

`python3.11 build_scripts/build_usd.py --no-examples --no-tutorials --no-tools --no-usdview --generator Ninja --verbose`

**Comment by sunyab** [comment]
Yes, USD should have the same thread-safety attributes with and without Python support. In the past we have seen and fixed threading issues related to the Python GIL. If you can share a stack trace during a hang there might be a clue there as to what's going on.

**Comment by ggarra13** [comment]
In my viewer I am using libbacktrace for stack traces which only gives me stack traces for the main thread unfortunately (do you happen to know how to improve this?), so I had to go to gdb.

Find attached a stack trace from Linux from gdb.  I Marked with "---------------" where the stack is hanging.  It seems to be an issue when calling pyLock.cpp.

[full_stack_trace.txt](https://github.com/user-attachments/files/15793651/full_stack_trace.txt)

Also, when I compiled with USD python support I also found issues with Wayland, but I'll open a new issue for that later, as it can wait.

**Comment by sunyab** [comment]
I'm not familiar with libbacktrace unfortunately, so I can't help you there.

Does your application interact with Python in any other way? From the stack trace it looks like USD is waiting on another thread to release the Python GIL, but I didn't see anything like that at a first glance.

**Comment by ggarra13** [comment]
My application has an embedded python created with pybind11, which gets started at main like:

```
int main(int argc, char* argv[])
{
    int r = 1;
#ifdef MRV2_PYBIND11
    // start the interpreter and keep it alive
    py::scoped_interpreter guard{};
#endif
    try
    {
        auto context = tl::system::Context::create();
        mrv::init(context);
        mrv::App app(argc, argv, context);
        r = app.getExit();
        if (0 == r)
        {
            r = app.run();
        }
    }
    catch (const std::exception& e)
    {
        std::cerr << "ERROR: " << e.what() << std::endl;
    }
    return r;
}
```

However, at the moment of loading an USD asset, there's no actual python code running, nor do I interact with the GIL in any way.
The big question is: why does Stage::Open need to run any python code at all when called from C++?  It is probably locking the GIL on a separate thread, which Python does NOT support.

**Comment by jesschimein** [comment]
Filed as internal issue #USD-9752

**Comment by sunyab** [comment]
Python supports locking the GIL in threads, but if one thread is holding on to the GIL while a second thread is trying to take it, the second thread will be forced to wait until the first releases it. In your case, pybind11's `scoped_interpreter` initializes and takes the GIL, so the worker threads trying to take the GIL are forced to wait. And since the main thread never releases the GIL, this turns into a hang.

This is outside the scope of OpenUSD, but there are a number of posts that discuss embedding a Python interpreter in a multi-threaded application that recommend unlocking the GIL in the main thread specifically to avoid issues like this. For example:

https://stackoverflow.com/questions/10625584/embedding-python-in-multithreaded-c-application

As to why UsdStage::Open runs Python at all -- when loading C++ libraries, USD will also load the corresponding Python modules (if any) if it detects that Python has been initialized. This ensures that all the necessary Python bindings are available if a C++ function that returns a USD type is called from Python.

**Comment by ggarra13** [comment]
@sunyab Thanks for your explanation.  I had figured as much.  One more question:  Does UsdStage::Open release the GIL once it is done with it or does it keep to itself?

**Comment by ggarra13** [comment]
@sunyab Never mind.  I just tried it and it works like a charm.  Thank you very much!

**Comment by sunyab** [comment]
Yes, the GIL will be released -- more specifically, the code that UsdStage::Open calls that ultimately tries to take the GIL will release it once it's done calling out to Python.

**Comment by sunyab** [comment]
Closing this out, but please let us know if you have more questions. Thanks!
